src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -1181,91 +1181,100 @@ void ShenandoahHeap::cancel_old_gc() {
-
-  // round down
-  size_t max_old_region_xfer = xfer_limit / region_size_bytes;
-
-  // We can limit the reserve to the size of anticipated promotions
-  size_t max_old_reserve = young_reserve * ShenandoahOldEvacRatioPercent / (100 - ShenandoahOldEvacRatioPercent);
-  // Here's the algebra:
-  //  TotalEvacuation = OldEvacuation + YoungEvacuation
-  //  OldEvacuation = TotalEvacuation*(ShenandoahOldEvacRatioPercent/100)
-  //  OldEvacuation = YoungEvacuation * (ShenandoahOldEvacRatioPercent/100)/(1 - ShenandoahOldEvacRatioPercent/100)
-  //  OldEvacuation = YoungEvacuation * ShenandoahOldEvacRatioPercent/(100 - ShenandoahOldEvacRatioPercent)
-
-  size_t reserve_for_mixed, reserve_for_promo;
-  if (doing_mixed) {
-    assert(old_generation()->available() >= old_generation()->free_unaffiliated_regions() * region_size_bytes,
-           "Unaffiliated available must be less than total available");
+  const size_t young_reserve = (young_generation()->max_capacity() * ShenandoahEvacReserve) / 100;
+  const size_t max_old_reserve = (ShenandoahOldEvacRatioPercent == 100) ?
+     old_available : MIN2((young_reserve * ShenandoahOldEvacRatioPercent) / (100 - ShenandoahOldEvacRatioPercent),
+                          old_available);
+
+  const size_t region_size_bytes = ShenandoahHeapRegion::region_size_bytes();
 
+  // Decide how much old space we should reserve for a mixed collection
+  size_t reserve_for_mixed = 0;
+  const size_t mixed_candidates = old_heuristics()->unprocessed_old_collection_candidates();
+  const bool doing_mixed = (mixed_candidates > 0);
+  if (doing_mixed) {
     // We want this much memory to be unfragmented in order to reliably evacuate old.  This is conservative because we
     // may not evacuate the entirety of unprocessed candidates in a single mixed evacuation.
     size_t max_evac_need = (size_t)
       (old_heuristics()->unprocessed_old_collection_candidates_live_memory() * ShenandoahOldEvacWaste);
+    assert(old_available >= old_generation()->free_unaffiliated_regions() * region_size_bytes,
+           "Unaffiliated available must be less than total available");
     size_t old_fragmented_available =
-      old_generation()->available() - old_generation()->free_unaffiliated_regions() * region_size_bytes;
+      old_available - old_generation()->free_unaffiliated_regions() * region_size_bytes;
     reserve_for_mixed = max_evac_need + old_fragmented_available;
     if (reserve_for_mixed > max_old_reserve) {
       reserve_for_mixed = max_old_reserve;
     }
-  } else {
-    reserve_for_mixed = 0;
   }
 
-  size_t available_for_promotions = max_old_reserve - reserve_for_mixed;
+  // Decide how much space we should reserve for promotions from young
+  size_t reserve_for_promo = 0;
+  const size_t promo_load = get_promotion_potential();
+  const bool doing_promotions = promo_load > 0;
   if (doing_promotions) {
     // We're only promoting and we have a maximum bound on the amount to be promoted
```

Comment:
With reference to the preceding comment, we should probably add an assert here: assert(!doing_mixed, "Don't do mixed and promotions in the same cycle")

Or maybe this is no longer true.  It doesn't have to be true.  The code that follows seems to suggest this is no longer true.  It was a simplification at some point in the implementation history.  If it is no longer true, let's remove that comment.

Reply:
I believe you are right. The comment appears obsolete. I'll check/confirm and delete the comment. Thanks!

Reply:
Done